Glsl.meta.slang raytracing #3560

jkwak-work opened the pull request:
This PR is only for a code review purpose.

ArielG-NV committed:
implementing ray tracing functions & ray_query extension tests

all functions part of ray_query, shader_invocation_reorder and ray_tracing_motion_blur are implemented for GLSL.meta. If HLSL lacks an implementation equal GLSL/SPIR-V was implemented without an HLSL equivalent in HLSL.meta (anticipation of HLSL equivalent function eventually); completed

built in variables have only been stubbed thus far; NOT completed

the following commit sets up formatting & tests for all ray_query functions (compute shader); completed

the commit also sets up structure for compile testing glsl raytracing shaders; NOT completed

ArielG-NV committed:
Implement all functions & built-in variables (with tests) for the GLSL_EXT_ray_query, GLSL_NV_shader_invocation_reorder & GLSL_NV_ray_tracing_motion_blur

Implement all functions & built-in variables (with tests) for the GLSL_EXT_ray_query, GLSL_NV_shader_invocation_reorder & GLSL_NV_ray_tracing_motion_blur glsl extension based on specification (https://github.com/KhronosGroup/GLSL/blob/main/extensions/nv/).

implementation details of worth{
All functions with alternatives in GLSL->HLSL have been implemented in GLSL, HLSL & SPRI-V (hlsl implementation has a few errors); otherwise only GLSL & SPIR-V was implemented.
* modified slang compiler to add support for query'ing and fetching raytracing payloads/attributes based on bound layout(location) and input a 'constexpr int'. This allows'dynamic typing' for these operations to late evaluate all location tied raytracing objects. An attribute to functions were also added to force setting a generic's type based on a payload object (this is also required to allow 'generic typed' raytracing object functions); all of the hacky generic forcing will be removed in next commit in regards to removing hlsl code

GL_EXT_ray_tracing was NOT implemented and part of scope of this commit, they will be for the next commit

This commit will be overrided with removing HLSL functions and logic. commit is happening just to keep code history

}

testing details of worth{
Raytracing shader (non compute, compute can test properly) tests only test compiling and opcodes due to testing system limitations
}

ArielG-NV committed:
Implement all functions & built-in variables (with tests) for the GLSL_EXT_ray_tracing, GLSL_EXT_ray_query, GLSL_NV_shader_invocation_reorder & GLSL_NV_ray_tracing_motion_blur glsl extension based on specification (https://github.com/KhronosGroup/GLSL/blob/main/extensions/).

implementation details of worth{
spirv required direct referencing of raytracing objects, as a result, __rayPayloadFromLocation, __rayAttributeFromLocation, and __rayCallableFromLocation, all take in a integer literal and use this to add a placeholder to fillin with an actual layout(location) bound object.

layout(location) bound raytracing objects are never optimized out, this was intentionally added to the code.

GL_EXT_ray_tracing was implemented as noted in the commit description.

all HLSL support has been removed.
}

testing details of worth{
Raytracing shader tests only test compiling and opcodes due to testing system limitations (raygen, intersection, miss, anyhit cover all raytracing functions/builtins/syntax); Compute raytracing works with actual value testing
}

ArielG-NV committed:
syntax adjustments

change dictionary usage to be more correct syntax-wise

compact and organize parser code to be more like other parser code (this structure is important since there will be many attributes along with layout() syntax later on)

code style adjustments to match slang use of braces

jkwak-work commented on tests/hlsl-intrinsic/shader-execution-reordering/hit-object-reorder-thread.slang:
```diff
@@ -74,14 +76,14 @@ void rayGenerationMain()
         someValues);
 
     uint r = calcValue(hit);
-    // SPIRV: OpReorderThreadWithHitObjectNV
```
Why is this test removed?

ArielG-NV replied:
I broke the tests during development; I forgot to add them back since I broke them at the start; I am adding them 👍

jkwak-work commented on tests/hlsl-intrinsic/shader-execution-reordering/hit-object-trace-ray.slang.expected:
Why is this file removed?

ArielG-NV replied:
these are not tests and go against the .gitignore; they don't do anything

jkwak-work commented on tests/glsl-intrinsic/raytracing/glsl-rayCompute.slang:
```diff
@@ -0,0 +1,106 @@
+    uint data[];
+} outputBuffer;
+
+__generic<T : __BuiltinFloatingPointType, let N : int, let M : int>
+bool equals(matrix<T, N, M> lhs, matrix<T, N, M> rhs)
+{
+    for (int i = 0; i < N; i++)
+    {
+        for (int j = 0; j < M; j++)
+        {
+            if (
+                lhs[i][j] 
+                !=
+                rhs[i][j]
+                )
+            {
+                return false;
+            }
+        }
+    }
+    return true;
+}
```
Don't we have a equality testing function already for Matrix?

ArielG-NV replied:
we do not, other testing code also has an "equals" function for this purpose

jkwak-work commented on source/slang/slang-ir.cpp:
```diff
@@ -4287,11 +4289,117 @@ namespace Slang
+            {
+                // should be impossible to hit
+                assert(false);
+            }
```
I think this can be better with
```
else
{
    SLANG_ASSERT(kIROp_SPIRVAsmOperandRayCallableFromLocation == op);
    varLayoutPointsTo = this->getRayCallableVariableFromLocation(intLitValue);
}

jkwak-work reviewed (commented):
It looks good to me overall.

ArielG-NV committed:
add back removed tests; adjust style as per requested with assert'ing impossible code

jkwak-work commented on source/slang/hlsl.meta.slang:
```diff
@@ -9531,22 +9531,32 @@ static const RAY_FLAG RAY_FLAG_SKIP_PROCEDURAL_PRIMITIVES       = 0x200;
 __target_intrinsic(hlsl, RayDesc)
 __target_intrinsic(cuda, RayDesc)
+//__target_intrinsic(glsl, RayDesc)
+//__target_intrinsic(spirv, RayDesc)
```
Can you either uncomment these or remove?

jkwak-work commented on source/slang/hlsl.meta.slang:
```diff
@@ -9822,7 +9832,8 @@ bool __reportIntersection(float tHit, uint hitKind)
     {
-    case _GL_EXT_ray_tracing: __intrinsic_asm "reportIntersectionEXT";
+    case _GL_EXT_ray_tracing:
+    case glsl: __intrinsic_asm "reportIntersectionEXT";
```
Is this change actually needed?
It seems like _GL_EXT_ray_tracing will cover all the GLSL target, doesn't it?

ArielG-NV replied:
glsl->_GL_EXT_ray_tracing does/did not seem to work, this is needed
I think its a superset->subset type of problem (_GL_EXT_ray_tracing -> glsl is fine, other way is not)

jkwak-work replied:
Do you mean you don't need to have "case _GL_EXT_ray_tracing" when you have "case glsl"?

ArielG-NV replied:
some code & tests enable/use the _GL_EXT_ray_tracing (hlsl->glsl) capability, I use glsl capability instead (but the code is identical since both are glsl)

jkwak-work commented on source/slang/hlsl.meta.slang:
```diff
@@ -9873,12 +9886,16 @@ void AcceptHitAndEndSearch()
+    case hlsl: __intrinsic_asm "DispatchRaysIndex";
+    case _GL_EXT_ray_tracing:
+    case glsl:
+    case glsl: 
```
This looks to be a mistake.
There should be only one "glsl" case.

jkwak-work commented on source/slang/hlsl.meta.slang:
```diff
@@ -11781,7 +12192,70 @@ struct HitObject
+        case spirv:
+            spirv_asm {
+                result:$$float = OpHitObjectGetCurrentTimeNV &this
+            };
```
Shouldn't this have "return"?
```
return spirv_asm { ... };
```

jkwak-work commented on source/slang/slang-emit-spirv.cpp:
```diff
@@ -2954,9 +2954,24 @@ struct SPIRVEmitContext
+            ensureExtensionDeclaration(UnownedStringSlice("SPV_KHR_ray_query"));
+            requireSPIRVCapability(SpvCapabilityRayQueryKHR);
+            goto rayExtensionAddition;
+rayExtensionAddition:;
```
Let's not use GOTO within switch statement.
In this case, you can set a temporary variable to be true and handle the action later.
```
bool isRayExtension = false;
...
case kIROp_VulkanCallablePayloadInDecoration:
...
    isRayExtension = true;
    break;
...
if (isRayExtension)
{
   emitOpDecorateLocation(...);
}
````
In general, "goto" should be limited to error handling; not for code reuse.

jkwak-work commented on source/slang/slang-ir.cpp:
```diff
@@ -5794,7 +5898,42 @@ namespace Slang
+    IRSPIRVAsmOperand* IRBuilder::emitSPIRVAsmOperandRayPayloadFromLocation(IRInst* inst)
+    {
+        SLANG_ASSERT(as<IRSPIRVAsm>(m_insertLoc.getParent()));
+        const auto i = createInst<IRSPIRVAsmOperand>(
```
It seems that you shouldn't have const here because your return type doesn't have const and "i" will be returned.

ArielG-NV replied:
const is used since other functions which do the same thing also set a const ptr for return; likely const is used since in the function we are signaling that the ptr's value won't change value between instantiation and returning

jkwak-work replied:
"const" is not a "signal"-ing keyword, and it cannot automatically become non-const.
Something doesn't look right on this code to me.

ArielG-NV replied:
I did some checking, in godbolt -- it should not compile in MSVC (without a flag), nor does it likely emit different ASM... I will check to see if there is a reason the other functions use a const~

ArielG-NV replied:
nope, just a cause of warnings; I suppose I will remove all of the const'ness from these type of functions and if there is a question of why that will be a different discussion

jkwak-work commented on source/slang/slang-ir.h:
```diff
@@ -2239,6 +2240,8 @@ struct IRAnalysis
     IRDominatorTree* getDominatorTree();
 };
 
+struct DiagnosticSink;
```
You may need to move this line up to the top where other classes are declared; around line 28~38.

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -7013,6 +7013,27 @@ namespace Slang
         {
             return SPIRVAsmOperand{ SPIRVAsmOperand::NonSemanticDebugPrintfExtSet, parser->ReadToken() };
         }
+        else if (AdvanceIf(parser, "__rayPayloadFromLocation")) {
```
You should keep the coding style same as how the previous code was written.
Please add a new line character between ")" and "{".
I see that this mistake occurs on multiple lines in this PR.

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -7771,6 +7793,20 @@ namespace Slang
         }
 
+#define CASE(key, type) if (AdvanceIf(parser, #key)) { auto modifier = parser->astBuilder->create<type>(); \
+    modifier->location = getIntegerLiteralValue(listBuilder.find<GLSLLayoutModifier>()->valToken); listBuilder.add(modifier); } else
```
Is "modifier->location" only for debugging purpose?

ArielG-NV replied:
modifier->location is for creating a modifier and tying a glsl layout(location) location value to it

jkwak-work reviewed (commented):
I have some comments.

ArielG-NV committed:
fixing code style & changing as per suggestions (9)

ArielG-NV committed:
remove useless const's which just add warnings

ArielG-NV committed:
Merge remote-tracking branch 'upstream/master' into glsl.meta.slang_raytracing_#3560

ArielG-NV committed:
changes to ensure branch works with upstream

fix various code-style warnings

fix vkray regression: remove assumption that layout-val will be attached to a vk object decoration (causes crashes in some specific senarios)

add missing ext's to hlsl.meta

removed __specialized_for_target(glsl) for specific function due to causing a regression and not doing anything functional

jkwak-work commented:
Looks good to me

ArielG-NV committed:
fix warning->error with int cast

ArielG-NV committed:
added capabilities to raytracing

fixed lots of coding style issues

added early intrinsic pass to handle spirv get_from_location and for the future 'glsl magic syntax'->'target with well defined syntax'

ArielG-NV committed:
add missing cap

ArielG-NV committed:
checkin after fixing all tests(?)

ArielG-NV committed:
fixed regressions by removing ext's when adding capabilities

(once capabilities auto-add ext's, these problems will be no more!)

ArielG-NV committed:
remove wrong rayIntersection shader code (fails compile causing failiure)

ArielG-NV committed:
push vsproj files

ArielG-NV committed:
Merge branch 'master' into glsl.meta.slang_raytracing_#3560

ArielG-NV committed:
push vsproj files

ArielG-NV committed:
adjust vs project filters

ArielG-NV committed:
realized I could just rerun the bat file...

ArielG-NV committed:
Merge branch 'master' into glsl.meta.slang_raytracing_#3560

csyonghe committed:
Merge branch 'master' into glsl.meta.slang_raytracing_#3560

jkwak-work committed:
Merge branch 'master' into glsl.meta.slang_raytracing_#3560

ArielG-NV committed:
Merge branch 'master' into glsl.meta.slang_raytracing_#3560

ArielG-NV committed:
adjusting syntax and code to conform with slang better

added missing cap's to raytracing & adjusted syntax+naming to be conformant with Slang

ArielG-NV committed:
Merge branch 'glsl.meta.slang_raytracing_#3560' of https://github.com/ArielG-NV/slang into glsl.meta.slang_raytracing_#3560

ArielG-NV committed:
fixes to remove allowGLSL branch of code for raytracing (allows hlsl syntax + glsl syntax for raytracing objects as well, which is quite cool)

changed comment about early raytracing object addition to ir (more about positioning of code emitted)

ArielG-NV committed:
add support for HLSL+GLSL raytracing (together)

change code emit'ing to ensure it emits properly at top of scope

refactored a bit of the descriptions to be more accurate for why we need things done the way they are

ArielG-NV committed:
Merge branch 'master' into glsl.meta.slang_raytracing_#3560

ArielG-NV committed:
change how we simplify SPIRV so it works when we have no references (adds to entry point)

ArielG-NV committed:
Merge branch 'glsl.meta.slang_raytracing_#3560' of https://github.com/ArielG-NV/slang into glsl.meta.slang_raytracing_#3560

ArielG-NV committed:
fix warning

ArielG-NV committed:
remove unused loop

ArielG-NV committed:
Merge branch 'master' into glsl.meta.slang_raytracing_#3560

ArielG-NV committed:
Merge branch 'master' into glsl.meta.slang_raytracing_#3560

ArielG-NV committed:
fix failing matrix buffer pointer

ArielG-NV committed:
minimize all scope of changes such that merge into master will have less conflicts

changed spirv replacement to be proper and not a hacky 'working' implementation (since the code is being finalized now and I don't need a 'just working' implementation for testing)

ArielG-NV committed:
fix forced emit'ing code hack due to finding source of problem being, no ref --> no HLSL export tag

ArielG-NV committed:
directly inline hanging call for clarity & styling

ArielG-NV committed:
Merge branch 'master' into glsl.meta.slang_raytracing_#3560
